printer/src/printer.rs

```diff
@@ -0,0 +1,334 @@
+    body: T,
+    annotation: Option<&Annotation>,
+    attributes: &Vec<Attribute>,
+    cb: F,
+  ) where
+    F: FnOnce(&mut Self, T),
+  {
+    self.push('{');
+    self.push(' ');
+
+    cb(self, body);
+
+    if let Some(annotation) = annotation {
+      if !matches!(self.out.chars().last(), Some(' ')) {
+        self.push(' ');
+      }
+
+      let Annotation::Function(fun) = annotation;
+      self.print_function(fun);
+    }
+
+    for attr in attributes {
+      self.print_attribute(attr);
+    }
+
+    self.push(' ');
+    self.push('}');
+  }
+
+  fn print_function(&mut self, fun: &Function) {
+    self.push(':');
+    self.print_identifier(&fun.id);
+
+    for option in &fun.options {
+      self.print_option(option);
+    }
+  }
+
+  fn print_identifier(&mut self, id: &Identifier) {
+    if let Some(namespace) = id.namespace {
+      self.push_str(namespace);
+      self.push(':');
+    }
+    self.push_str(id.name);
+  }
+
+  fn print_option(&mut self, option: &FnOrMarkupOption) {
+    self.push(' ');
+    self.print_identifier(&option.key);
+    self.push('=');
+    dispatch_print!(self, &option.value, LiteralOrVariable {
+      Literal => print_literal,
+      Variable => print_variable,
+    })
+  }
+
+  fn print_literal(&mut self, lit: &Literal) {
+    dispatch_print!(self, lit, Literal {
+      Text => print_text,
+      Quoted => print_quoted,
+      Number => print_number,
+    });
+  }
+
+  fn print_quoted(&mut self, quoted: &Quoted) {
+    self.push('|');
+    for part in &quoted.parts {
+      dispatch_print!(self, part, QuotedPart {
+        Text => print_text,
+        Escape => print_escape,
+      });
+    }
+    self.push('|');
+  }
+
+  fn print_number(&mut self, num: &Number) {
+    self.push_str(num.raw);
+  }
+
+  fn print_variable(&mut self, var: &Variable) {
+    self.push('$');
+    self.push_str(var.name);
+  }
+
+  fn print_attribute(&mut self, attr: &Attribute) {
+    self.push(' ');
+    self.push('@');
+    self.print_identifier(&attr.key);
+
+    if let Some(value) = &attr.value {
+      self.push('=');
+      self.print_literal(value);
+    }
+  }
+
+  fn print_markup(&mut self, markup: &Markup) {
+    self.push('{');
+    if let MarkupKind::Close = markup.kind {
+      self.push('/');
+    } else {
+      self.push('#');
+    }
+
+    self.print_identifier(&markup.id);
+
+    for option in &markup.options {
+      self.print_option(option);
+    }
+    for attr in &markup.attributes {
+      self.print_attribute(attr);
+    }
+
+    self.push(' ');
+    if let MarkupKind::Standalone = markup.kind {
+      self.push('/');
+    }
+    self.push('}');
+  }
+
+  fn print_complex_message(&mut self, message: &ComplexMessage) {
+    let mut is_input = None;
+
+    for decl in &message.declarations {
+      let prev_is_input =
+        is_input.replace(matches!(decl, Declaration::InputDeclaration(_)));
+      if prev_is_input.is_some() && prev_is_input != is_input {
+        self.push('\n');
+      }
```

Comment:
I don't like these new lines that separate inputs, locals, and the body. They will often be weird, because you may have a input and a local that logically fit in a group, but will now be artificially separated and instead grouped with unrelated locals or inputs.